core/src/avm2/call_stack.rs

```diff
@@ -0,0 +1,89 @@
+use crate::avm2::method::{MethodKind, MethodMetadata, MethodPosition};
+use crate::string::WString;
+use gc_arena::{Collect, MutationContext};
+
+#[derive(Collect, Debug, Clone, Copy)]
+#[collect(no_drop)]
+pub struct CallStackNode<'gc> {
+    meta: Option<MethodMetadata<'gc>>,
+    method_id: Option<u32>,
+}
+
+impl<'gc> CallStackNode<'gc> {
+    /// Creates an empty call stack node. This should only be used for nothing activations.
+    pub fn empty() -> Self {
+        Self {
+            meta: None,
+            method_id: None,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.meta.is_none() && self.method_id.is_none()
+    }
+
+    /// Helper method for correctly constructing script init call nodes.
+    pub fn script_init() -> Self {
+        Self {
+            meta: Some(MethodMetadata::new_script_init()),
+            method_id: None,
+        }
+    }
+
+    /// Creates a new CallStackNode.
+    ///
+    /// It is an error for both `meta` and `method_id` to be None at the same time.
+    pub fn new(meta: Option<MethodMetadata<'gc>>, method_id: Option<u32>) -> Self {
```

Comment:
Could this parameter be an enum instead of two options? Even if it could possibly have both, maybe one variant can have `Metadata+method_id`, while the other only has `method_id`.

Reply:
I decided to just make CallStackNode itself an enum, because while `meta` and `method_id` could be available at the same time, we never actually use `method_id` if `meta` is available anyways.